### PR TITLE
fix(ISD-362): Remove unused resources

### LINKED ISSUE #4: Leaked Kubernetes resources when a service-name / service-hostname changes
The nginx-ingress-integrator charm creates Kubernetes resources (Services and Ingress Resources) based on the configured service-name and service-hostname.

However, if the charm is then configured with a different service-name and service-hostname, the previous resources are not removed. This happens everytime the configuration options change since the Resource names are based on the aforementioned service-name / service-hostname.

This can also happen for the service-names and service-hostnames received from relation data (+ additional-hostnames for Ingress Resources). This issue occurs on previous revisions as well.

Because the Kubernetes resource names are based on service-name / service-hostname, cleaning them up is not immediately straight-forward since the charm would have to store the previous values and compare them to the current ones.

### PR DESCRIPTION
**Pull request by amandahla**
Fix #4 

This PR:
- Add a mechanism that for every event, checks if there are unused services or ingress and delete them. 
This is done by:
1) getting all services/ingress resources with the label "app.juju.is/created-by=".
2) getting all services/ingress resources from config or relations 
3) Each service/ingress resource that is not in the list mentioned above is deleted.
- Add unit and integration tests
- Add mypy checker and related fixes or disabled types
- Change AnyCharm to get Ingress values from the relation, if exists. The default values were set in the _\_init_\_ and since charm is a stateless application, every event was initializing the charm and recreating the ingress with the default values. Reference [here](https://juju.is/docs/sdk/create-a-minimal-kubernetes-charm#heading--define-the-charm-initialisation-and-application-services).

Findings that will be addressed in further PRs:
- Custom label for NGINX charm (to be confirmed)
- Refactoring of unit tests (ISD-417)
- Refactoring of Type hints (ISD-422)

### CODE REVIEW

File: src/charm.py [hunk 1]
```diff
@@ -467,6 +475,42 @@ def _has_required_fields(self, conf_or_rel: _ConfigOrRelation):
...
+        all_services = api.list_namespaced_service(
+            namespace=self._namespace, label_selector=created_by_label
+        )
+        all_svc_names = [item.metadata.name for item in all_services.items]
+        unused_svc_names = [
+            svc_name
+            for svc_name in all_svc_names
+            if svc_name.replace("-service", "") not in current_svc_names
+        ]
+        LOGGER.info(
```
Comment (gregory-schiano) [inline src/charm.py:497]: I think this one should be debug

File: src/charm.py [hunk 2]
```diff
@@ -550,6 +594,31 @@ def _look_up_and_set_ingress_class(self, api, body):
...
+        api = _networking_v1_api()
+        created_by_label = f"{CREATED_BY_LABEL}{self.app.name}"
+        all_ingresses = api.list_namespaced_ingress(
+            namespace=self._namespace, label_selector=created_by_label
+        )
+        all_svc_hostnames = [ingress.spec.rules[0].host for ingress in all_ingresses.items]
+        unused_svc_hostnames = [
+            hostname for hostname in all_svc_hostnames if hostname not in current_svc_hostnames
+        ]
+        LOGGER.info(
```
Comment (gregory-schiano) [inline src/charm.py:613]: Same as above

File: src/charm.py [hunk 3]
```diff
@@ -23,6 +23,7 @@
 LOGGER = logging.getLogger(__name__)
 _INGRESS_SUB_REGEX = re.compile("[^0-9a-zA-Z]")
 BOOLEAN_CONFIG_FIELDS = ["rewrite-enabled"]
+CREATED_BY_LABEL = "app.juju.is/created-by="
```
Comment (gregory-schiano) [inline src/charm.py:26]: As discussed in the daily digest, I don't understand where this label comes from.
The charm uses python kubernetes-client library and doesn't explicitly set this label which means:
- Either the library takes it from the pod calling the API (I don't think that's the case, I quickly took a look at the library code and didn't find such business logic)
- Or it's added by Juju for every ressource within the namespace (perhaps edited automatically by the modelcontroller ? But how does juju show who created this kubernetes ressources as there could be multiple charms deployed in a model)

This should be clarified for 2 reasons:
- Ensure we won't delete services/ingress we don't want to delete
- Ensure this label won't be removed in a future releases

One approach could be to add our own label in this charm when creating the Ingress/Service, the drawback is that we won't be able to cleanup previously created ressources (in case of an upgrade of charm)

Comment (amandahla) [reply src/charm.py:26]: After some digging, I think I found out how Juju knows which application created the resource.

When Juju receives a request via cassadmission, fetches the juju application name associated with a service account that in its turn has this value set in the label "app.kubernetes.io/name".
 
Code:
https://github.com/juju/juju/blob/develop/worker/caasadmission/handler.go#L131

microk8s output:
```
$ microk8s kubectl get serviceaccount -n indico nginx1 --show-labels
NAME     SECRETS   AGE     LABELS
nginx1   0         4h48m   app.kubernetes.io/managed-by=juju,app.kubernetes.io/name=nginx1
$ microk8s kubectl get ing -n indico --show-labels
NAME                    CLASS    HOSTS           ADDRESS     PORTS   AGE     LABELS
indico-local-ingress    public   indico.local    127.0.0.1   80      4h46m   app.juju.is/created-by=nginx1
hello-kubecon-ingress   public   hello-kubecon   127.0.0.1   80      6m5s    app.juju.is/created-by=nginx2
```

File: src/charm.py [hunk 4]
```diff
@@ -299,13 +313,7 @@ def _get_k8s_ingress(self):
...
 
         hostnames = [self._service_hostname]
-        hostnames.extend(
-            [
-                x
-                for x in self._get_config_or_relation_data("additional-hostnames", "").split(",")
-                if x
-            ]
-        )
+        hostnames.extend(self._additional_hostnames)
```
[the remark this answers is not in the document]
Comment (amandahla) [reply src/charm.py:316]: No, it's used in _on_config_changed too.

File: src/charm.py [hunk 5]
```diff
@@ -114,6 +115,19 @@ def _get_relation_data_or_config(self, field, fallback):
...
         return fallback
 
+    @property
+    def _additional_hostnames(self) -> List[str]:
+        """Return a list with additional hostnames.
+
+        Returns:
+            The additional hostnames set by configuration already split by comma.
+        """
+        return [
```
[the remark this answers is not in the document]
Comment (amandahla) [reply src/charm.py:125]: Thanks for pointing that out but since one of the use cases is to check if a hostname is in the list (current hostname + all additional hostnames), an iterator wouldn't be a good fit.

File: src/charm.py [hunk 6]
```diff
@@ -467,6 +475,42 @@ def _has_required_fields(self, conf_or_rel: _ConfigOrRelation):
...
+            current_svc_names: service names set by config or relation data
+            current_svc_hostnames: service hostnames set by config or relation data
+        """
+        self.k8s_auth()
+        api = _core_v1_api()
+        created_by_label = f"{CREATED_BY_LABEL}{self.app.name}"
+        all_services = api.list_namespaced_service(
+            namespace=self._namespace, label_selector=created_by_label
+        )
+        all_svc_names = [item.metadata.name for item in all_services.items]
```
[the remark this answers is not in the document]
Comment (amandahla) [reply src/charm.py:491]: Do you think would be worth using a generator instead of a list comprehension for a small list? I'm assuming that would be rare to be more than 100 items or even 50 in this case.

File: src/charm.py [hunk 7]
```diff
@@ -550,6 +594,31 @@ def _look_up_and_set_ingress_class(self, api, body):
...
 
+    def _delete_unused_ingresses(self, current_svc_hostnames: List[str]):
+        """Delete ingresses that are no longer managed by nginx-ingress-integrator charm.
+
+        Args:
+            current_svc_hostnames: service hostnames set by config or relation data
+        """
+        self.k8s_auth()
+        api = _networking_v1_api()
+        created_by_label = f"{CREATED_BY_LABEL}{self.app.name}"
```
[the remark this answers is not in the document]
Comment (amandahla) [reply src/charm.py:605]: They use different APIs :(

File: src/charm.py [hunk 8]
```diff
@@ -745,6 +813,17 @@ def _on_config_changed(self, _):
                     msgs.append(f"Ingress IP(s): {', '.join(ingress_ips)}")
                 msgs.append(f"Service IP(s): {', '.join(self._report_service_ips())}")
                 msg = ", ".join(msgs)
+                self._delete_unused_services(svc_names)
+                svc_hostnames = [
```
[the remark this answers is not in the document]
Comment (amandahla) [reply src/charm.py:817]: They are separated because in this way one function does just one thing and would be easier to modify/debug if needed. Another reason is that they use different API calls.

File: src/charm.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (amandahla) [reply src/charm.py]: (I'm asking out of curiosity, I applied your suggestion)

File: tests/integration/test_relation.py [hunk 9]
```diff
@@ -51,6 +53,38 @@ async def build_and_deploy_ingress():
...
+    model_name = ops_test.model.name
+    created_by_label = f"{CREATED_BY_LABEL}ingress"
+    all_ingresses = api_networking.list_namespaced_ingress(
+        namespace=model_name, label_selector=created_by_label
+    )
+    all_svc_hostnames = [ingress.spec.rules[0].host for ingress in all_ingresses.items]
+    assert all_svc_hostnames == ["any"]
+    await ops_test.juju("config", INGRESS_APP_NAME, "service-hostname=new-name")
+    await ops_test.model.wait_for_idle(status="active")
+    all_ingresses = api_networking.list_namespaced_ingress(
```
[the remark this answers is not in the document]
Comment (amandahla) [reply tests/integration/test_relation.py]: Good catch.

File: tests/unit/test_charm.py [hunk 10]
```diff
@@ -20,6 +20,8 @@ def setUp(self):
         self.addCleanup(self.harness.cleanup)
         self.harness.begin()
 
+    @patch("charm.NginxIngressCharm._delete_unused_ingresses")
```
[the remark this answers is not in the document]
Comment (amandahla) [reply tests/unit/test_charm.py:23]: Thanks for sharing the blog post. Since would be out of the scope of this PR, I added the autospec parameter only to the mocks created by this PR. I believe we could open a new story for refactoring the tests, what do you think?

File: src/charm.py (hunk 3 above)
Comment (gregory-schiano) [reply src/charm.py:29]: Ah makes sens, it uses an Admission Controller to get the events, and uses the Service Account attached to the pod.
Thx for digging, good to know anyway :)

Did you get some feedback from Juju team on this topic ?

File: src/charm.py [hunk 11]
```diff
@@ -123,9 +123,11 @@ def _additional_hostnames(self) -> List[str]:
             The additional hostnames set by configuration already split by comma.
         """
         return [
-            x
-            for x in self._get_config_or_relation_data("additional-hostnames", "").split(",")
-            if x
+            hostname
+            for hostname in self._get_config_or_relation_data("additional-hostnames", "").split(
+                ","
+            )
```
Comment (gregory-schiano) [inline src/charm.py:132]: The indentation seems strange, wouldn't it be better to return to the line before ".split" ?

File: src/charm.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (jdkandersson) [reply src/charm.py:128]: That is on the user side, it doesn't stop you from returning an iterator and the caller passing it to the list constructor. By returning a list, the use of a list is forced rather than up to the caller. The caller could be just interested in the first one or just need to process it

File: src/charm.py [hunk 12]
```diff
@@ -467,6 +475,42 @@ def _has_required_fields(self, conf_or_rel: _ConfigOrRelation):
         field_names = [f'_{f.replace("-", "_")}' for f in REQUIRED_INGRESS_RELATION_FIELDS]
         return all(getattr(conf_or_rel, f) for f in field_names)
 
+    def _delete_unused_services(self, current_svc_names: List[str]):
```
[the remark this answers is not in the document]
Comment (jdkandersson) [reply src/charm.py]: Was the mypy configuration added? I can't spot the change in this PR?

File: src/charm.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (jdkandersson) [reply src/charm.py]: If a generator works, why not use it? It is more efficient (has less functionality than a list) and uses less memory

File: src/charm.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (jdkandersson) [reply src/charm.py:610]: Take the API is a parameter?

File: src/charm.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (jdkandersson) [reply src/charm.py:822]: What about adding another function that removes all the unused resources rather than writing that in the `_on_config_changed` code?

File: tests/integration/test_relation.py [hunk 13]
```diff
@@ -51,6 +53,38 @@ async def build_and_deploy_ingress():
...
 
 
+@pytest.mark.usefixtures("build_and_deploy")
+async def test_delete_unused_ingresses(ops_test: OpsTest):
+    """
+    arrange: given charm has been built, deployed and related to a dependent application
+    act: when the service-hostname is changed and when is back to previous value
+    assert: then the workload status is active and the unused ingress is deleted
+    """
+    kubernetes.config.load_kube_config()
```
[the remark this answers is not in the document]
Comment (jdkandersson) [reply tests/integration/test_relation.py:64]: I meant in the code itself, i.e., after the block of code that does the arrange, add an empty line. Looks a bit better now anyway

File: tests/unit/test_charm.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (jdkandersson) [reply tests/unit/test_charm.py:23]: Yeh it is probably reasonable to do so, although this is kind of digging a deeper hole. The use of mocks her looks excessive which could indicate that the unit tests should actually be integration tests or that we need to write a few abstractions for the external dependencies that are easier to mock. Any thoughts @gregory-schiano?

[the remark this answers is not in the document]
Comment (gregory-schiano) [reply tests/unit/test_charm.py:23]: Yeap perhaps doing like Weii did on Spring Boot Charm, creating some mock classes 
https://github.com/canonical/spring-boot-k8s-operator/blob/main/tests/unit/spring_boot_patch.py

Globally speaking if we start needing more mock classes like this, it could be helpful to make a "Charm Mock module" in a separate repo we could easily import and use instead of defining those in multiple charms

File: src/charm.py (hunk 3 above)
Comment (amandahla) [reply src/charm.py:29]: Yes, here it is:
https://chat.charmhub.io/charmhub/pl/3ik6j6ryrjby7ykjx8fmy9keha

```
I can't guarantee these won't change in the future. We'll try to ensure they don't get changed, but I would recommend implementing a test on your end.

Why I could see this changing at some point is two things, labelling of resources created by the app isn't perfect, I think there is still some edge cases to work through; secondly the app name isn't unique, so we need to add a model-uuid into it, since resources could be created outside the model managed namespace.
```

File: src/charm.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (amandahla) [reply src/charm.py]: Sorry, I just added the "-> None". Next commit will have mypy.

File: src/charm.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (amandahla) [reply src/charm.py:822]: Done.

File: src/charm.py (hunk 11 above)
Comment (amandahla) [reply src/charm.py:132]: Sure, changed. :)

File: src/charm.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (amandahla) [reply src/charm.py]: BTW I added a lot of "type: ignore"/"disable"s but without comment because would be repetitive. I'm not sure if this is the best approach.
There are many cases like:
```
tests/integration/conftest.py:40: error: Item "None" of "Optional[Model]" has no attribute "deploy"  [union-attr]
```

File: src/charm.py (hunk 3 above)
Comment (amandahla) [reply src/charm.py:31]: We could work with both options. First, we could look for our label. If doesn't exist then we look for the app.juju.is/created-by=. If none of them exist, then nothing will be deleted. What do you think?

File: src/charm.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (amandahla) [reply src/charm.py]: Changed :)

File: src/charm.py [hunk 14]
```diff
@@ -745,6 +813,17 @@ def _on_config_changed(self, _):
...
                 msg = ", ".join(msgs)
+                self._delete_unused_services(svc_names)
+                svc_hostnames = [
+                    conf_or_rel._service_hostname for conf_or_rel in self._all_config_or_relations
+                ]
+                all_additional_hostnames = [
+                    conf_or_rel._additional_hostnames
+                    for conf_or_rel in self._all_config_or_relations
+                ]
+                for additional_hostname in all_additional_hostnames:
```
[the remark this answers is not in the document]
Comment (amandahla) [reply src/charm.py]: Moved to self._delete_unused_resources()

File: src/charm.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (jdkandersson) [reply src/charm.py]: So for those just put an `assert app.model` or whatever at the start of the test and they will go away. Mypy is telling you that `ops` doesn't guarantee the model is defined - therefore we should check that it is if we are going to use it

File: pyproject.toml [hunk 15]
```diff
@@ -14,6 +14,9 @@ branch = true
 fail_under = 93
 show_missing = true
 
+[tool.mypy]
```
Comment (jdkandersson) [inline pyproject.toml:17]: Specifically, the contributing guide asks for this:

```tox
[tool.mypy]
check_untyped_defs = true
disallow_untyped_defs = true

[[tool.mypy.overrides]]
module = "tests.*"
disallow_untyped_defs = false
```

File: src/charm.py [hunk 16]
```diff
@@ -116,17 +123,16 @@ def _get_relation_data_or_config(self, field, fallback):
...
         Returns:
             The additional hostnames set by configuration already split by comma.
         """
-        return [
-            x
-            for x in self._get_config_or_relation_data("additional-hostnames", "").split(",")
-            if x
-        ]
+        additional_hostnames = self._get_config_or_relation_data("additional-hostnames", "")
+        for hostname in additional_hostnames.split(","):
```
Comment (jdkandersson) [inline src/charm.py:133]: You could do this by writing: `yield from filter(None, additional_hostnames.split(","))`

File: src/charm.py [hunk 17]
```diff
@@ -439,25 +443,40 @@ def k8s_auth(self):
...
+            return _networking_v1_api()
+        LOGGER.error("k8s api %s not found", api)
+        return None
+
     def _report_service_ips(self) -> List[str]:
         """Report on service IP(s) and return a list of them."""
-        self.k8s_auth()
-        api = _core_v1_api()
-        services = api.list_namespaced_service(namespace=self._namespace)
+        api = self.k8s_api(KUBERNETES_CORE_API)
```
Comment (jdkandersson) [inline src/charm.py:462]: Is this an improvement? Maybe we have a function `core_v1_api` that returns the `core_v1_api` after calling `k8s_auth`? We probably lose a lot of type hint value by making the api to use generic

File: src/charm.py [hunk 18]
```diff
@@ -699,7 +719,12 @@ def _process_ingresses(self, ingresses):
                 # Ensure that the exact same route for this route was not yet defined.
                 defined_paths = ingress_paths[rule.host]
                 for path in rule.http.paths:
-                    duplicate_paths = list(filter(lambda p: p.path == path.path, defined_paths))
+                    duplicate_paths = list(
+                        filter(
+                            lambda p: p.path == path.path,  # pylint: disable=cell-var-from-loop
```
Comment (jdkandersson) [inline src/charm.py:724]: SHould we note why this disable is reasonable in this case please?

File: src/charm.py [hunk 19]
```diff
@@ -813,17 +850,7 @@ def _on_config_changed(self, _):
...
-                    conf_or_rel._service_hostname for conf_or_rel in self._all_config_or_relations
-                ]
-                all_additional_hostnames = [
-                    conf_or_rel._additional_hostnames
-                    for conf_or_rel in self._all_config_or_relations
-                ]
-                for additional_hostname in all_additional_hostnames:
-                    svc_hostnames.extend(additional_hostname)
-                self._delete_unused_ingresses(svc_hostnames)
+                self._delete_unused_resources()
```
Comment (jdkandersson) [inline src/charm.py:853]: Nice!

File: src/charm.py [hunk 20]
```diff
@@ -699,7 +719,12 @@ def _process_ingresses(self, ingresses):
                 # Ensure that the exact same route for this route was not yet defined.
                 defined_paths = ingress_paths[rule.host]
                 for path in rule.http.paths:
-                    duplicate_paths = list(filter(lambda p: p.path == path.path, defined_paths))
+                    duplicate_paths = list(
```
Comment (jdkandersson) [inline src/charm.py:722]: This function looks like it would benefit from refactoring, there are a lot of for loops in for loops, if statement etc

File: src/charm.py (hunk 18 above)
Comment (jdkandersson) [reply src/charm.py:724]: This can also be done without a lambda: `duplicate_paths = [defined_path for defined_path in defined_paths if defined_path.path == path.path]`

Writing this I'm confused by all the paths, the variable names here could be improved for both `defined_paths` and `path` since both of them have a `path` attribute which means that the outer variable probably is more than just a `path`

File: src/charm.py (hunk 16 above)
Comment (jdkandersson) [reply src/charm.py:133]: Or yield from `(hostname for hostname in additional_hostnames if hostname)` depending on preference

File: src/charm.py [hunk 21]
```diff
@@ -431,25 +443,40 @@ def k8s_auth(self):
 
         self._authed = True
 
+    def k8s_api(self, api: str) -> object:
```
Comment (jdkandersson) [inline src/charm.py:446]: I'm curios what the reason was for adding this?

File: src/charm.py (hunk 18 above)
Comment (merkata) [reply src/charm.py:724]: +1 for rephrasing, all this "path path, path" is rhythmic, though it's better suited in a song and not in code.

File: src/charm.py (hunk 3 above)
Comment (gregory-schiano) [reply src/charm.py:33]: Well not sure the added business logic worth it
- Having our own label, will never trigger the fallback to "created-by", unless someone removes it from our code. In that case we should ensure a unit/integration test will fail due to broken feature
- Relying on juju "created-by" label, will only break if they remove it, and here once more we should have a unit/integration tests that will fail if that's the case.

I'm more in favor of the 1st one because we don't run our integration tests scheduled, which means that the feature could be broken without us noticing until we do a code change that breaks the integration tests.

File: src/charm.py (hunk 17 above)
Comment (amandahla) [reply src/charm.py:462]: I might misunderstand one of your reviews when you said "Take the API is a parameter?" so I created this method receiving the wanted API as a parameter. I'll review if it's ok to change every call to core_v1_api/networking to include self.k8s_auth instead.

File: src/charm.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (amandahla) [reply src/charm.py]: Thanks!

File: src/charm.py (hunk 16 above)
Comment (amandahla) [reply src/charm.py]: Changed, thanks

File: src/charm.py (hunk 18 above)
Comment (amandahla) [reply src/charm.py]: Changed.

File: src/charm.py (hunk 21 above)
Comment (amandahla) [reply src/charm.py]: Removed :)

File: tests/integration/any_charm.py [hunk 22]
```diff
@@ -8,9 +8,12 @@
 import pathlib
 import signal
 import subprocess
+from typing import Dict
 
-from any_charm_base import AnyCharmBase
-from ingress import IngressRequires
+from any_charm_base import AnyCharmBase  # type: ignore[import]
```
Comment (jdkandersson) [inline tests/integration/any_charm.py:13]: I think these can be avoided by adding the dependencies to the `lint` job that `mypy` is looking for

File: tests/integration/any_charm.py [hunk 23]
```diff
@@ -41,6 +44,49 @@ def update_ingress(self, ingress_config):
...
+    def _has_required_fields(self, rel: Dict) -> bool:
+        """Check for required fields in relation
+
+        Args:
+            rel: relation to check
+
+        Returns:
+            Returns true if all fields exist
+        """
+        return "service-hostname" in rel and "service-name" in rel and "service-port" in rel
```
Comment (jdkandersson) [inline tests/integration/any_charm.py:56]: How about `return all(key in rel for key in ("service-hostname", "service-name", "service-port"))`?

File: tests/integration/any_charm.py [hunk 24]
```diff
@@ -41,6 +44,49 @@ def update_ingress(self, ingress_config):
...
+        """Get ingress config from relation or default
+
+        Returns:
+            The ingress config to be used
+        """
+        if self._has_ingress_relation() and self._has_app_data():
+            rel = self.model.relations["ingress"][0].data[self.app]
+            if self._has_required_fields(rel):
+                return {
+                    "service-hostname": rel["service-hostname"],
```
Comment (jdkandersson) [inline tests/integration/any_charm.py:84]: Would it be worth making these constants? They seem to be used in a few places

File: tests/integration/conftest.py [hunk 25]
```diff
@@ -10,7 +10,7 @@
 from pathlib import Path
 from typing import List
 
-import kubernetes
+import kubernetes  # type: ignore[import]
```
Comment (jdkandersson) [inline tests/integration/conftest.py:13]: Similar to earlier, this could potentially be resolved if `kubernetes` is added to the `deps` of the `[testenv:lint]` environment in `tox.ini`

File: tests/integration/test_relation.py [hunk 26]
```diff
@@ -43,7 +43,7 @@ async def build_and_deploy(ops_test: OpsTest, run_action):
 
     async def build_and_deploy_ingress():
         charm = await ops_test.build_charm(".")
-        return await ops_test.model.deploy(
+        return await ops_test.model.deploy(  # type: ignore[union-attr]
```
Comment (jdkandersson) [inline tests/integration/test_relation.py:46]: How about adding an `assert ops_test.model` somewhere earlier so that `ops_test.model` is guaranteed not to be `None` in the test?

File: tests/unit/test_charm.py [hunk 27]
```diff
@@ -1,13 +1,13 @@
 # Copyright 2023 Canonical Ltd.
 # See LICENSE file for licensing details.
 
-
 import unittest
+from typing import List
```
Comment (jdkandersson) [inline tests/unit/test_charm.py:5]: Which Python version is being targeted by the charm? If it is 3.10 or later `list` can be used instead of `typing.List` which is deprecated

File: tests/unit/test_library_ingress.py [hunk 28]
```diff
@@ -60,7 +61,7 @@ class CharmWithConfigDict(NginxIngressConsumerCharm):
             def __init__(self, *args):
                 super().__init__(*args, config_dict=config_dict)
 
-        self.harness = Harness(CharmWithConfigDict, meta=META)
+        self.harness = Harness(CharmWithConfigDict, meta=META)  # type: ignore[arg-type]
```
Comment (jdkandersson) [inline tests/unit/test_library_ingress.py:64]: Is it worth describing the problem and why it is reasonable to ignore?

Comment (jdkandersson) [reply tests/unit/test_library_ingress.py:64]: Also applies for the other disable rules, see here for the standard: https://github.com/canonical/is-charms-contributing-guide/blob/main/CONTRIBUTING.md#static-code-analysis

File: src/charm.py [hunk 29]
```diff
@@ -227,15 +235,15 @@ def _service_name(self):
         return self._get_config_or_relation_data("service-name", "")
 
     @property
-    def _service_port(self):
+    def _service_port(self) -> Any:
```
Comment (jdkandersson) [inline src/charm.py:238]: Isn't it `int`? `Any` is kind of pointless since it essentially disables typing checks, perhaps we make a follow up story to clean up the type hints?

File: src/charm.py [hunk 30]
```diff
@@ -431,25 +431,36 @@ def k8s_auth(self):
 
         self._authed = True
 
+    def _core_v1_api(self) -> Any:
```
Comment (jdkandersson) [inline src/charm.py:434]: Nice! Wouldn't the type be `CoreV1Api`?

File: src/charm.py [hunk 31]
```diff
@@ -431,25 +431,36 @@ def k8s_auth(self):
...
+        """Use the v1 k8s API."""
+        self.k8s_auth()
+        return kubernetes.client.CoreV1Api()
+
+    def _networking_v1_api(self) -> Any:
```
Comment (jdkandersson) [inline src/charm.py:439]: Wouldn't the type be `NetworkingV1Api`? I guess for the functions add/ changed in the PR it would be good to write type hints that aren't `Any`. For the rest, I can see an argument for that it should be done in a follow up PR

File: tests/unit/test_charm.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (amandahla) [reply tests/unit/test_charm.py:24]: ISD-417 was created to address this.

File: tests/integration/any_charm.py (hunk 22 above)
Comment (amandahla) [reply tests/integration/any_charm.py:13]: But this file is used inside the Any charm that has this dependency. any_charm_base is not used by the charm itself.

File: tests/integration/conftest.py (hunk 25 above)
Comment (amandahla) [reply tests/integration/conftest.py:13]: Without this, I get the error "error: Skipping analyzing "kubernetes": module is installed, but missing library stubs or py.typed marker  [import]". 
It seems that kubernetes doesn't have type stubs, if I understood correctly. 
We might consider using the plugin https://pypi.org/project/kubernetes-typed/ in a further PR.

File: tests/unit/test_library_ingress.py (hunk 28 above)
Comment (amandahla) [reply tests/unit/test_library_ingress.py:64]: I believe to fix this one would be necessary to change CharmType as covariant = True in the Harness class

https://mypy.readthedocs.io/en/stable/generics.html#variance-of-generic-types
https://ops.readthedocs.io/en/latest/_modules/ops/testing.html#Harness

File: src/charm.py (hunk 29 above)
Comment (amandahla) [reply src/charm.py:238]: My bad, is int, thanks.
A follow up story would be great also considering the plugin to check for Kubernetes types.

File: src/charm.py (hunk 31 above)
Comment (amandahla) [reply src/charm.py:439]: Yes, it would be nice to have a closer look at this together with the plugin for Kubernetes types + mypy.

File: tests/integration/any_charm.py (hunk 24 above)
Comment (amandahla) [reply tests/integration/any_charm.py]: Nice :)

File: tests/integration/test_relation.py (hunk 26 above)
Comment (amandahla) [reply tests/integration/test_relation.py]: My first attempt didn't work out but now it's ok, thanks.

File: tests/unit/test_library_ingress.py (hunk 28 above)
Comment (amandahla) [reply tests/unit/test_library_ingress.py:65]: Comment added.

File: src/charm.py (hunk 29 above)
Comment (amandahla) [reply src/charm.py]: ISD-422 was created to address type hints.

File: tests/integration/conftest.py (hunk 25 above)
Comment (amandahla) [reply tests/integration/conftest.py:13]: ISD-422 was created to address type hints.

File: tests/integration/any_charm.py (hunk 22 above)
Comment (jdkandersson) [reply tests/integration/any_charm.py:13]: Oh I see, it is something that would be added in the any charm context, so it wouldn't be possible to add this as a pip dependency

**Review by jdkandersson — APPROVED** [review]

**Review by merkata — APPROVED** [review]
Looks great!

File: tests/integration/any_charm.py (hunk 22 above)
Comment (amandahla) [reply tests/integration/any_charm.py:13]: Exactly :) I'll mark this as resolved but this can be reviewed in the story ISD-422.

File: src/charm.py (hunk 3 above)
Comment (amandahla) [reply src/charm.py:31]: We could work on this in a new story if it's ok for you. And for this one, we continue using the Juju label.
